### ISSUE TITLE
Currently impossible to get a user's own account

### ISSUE THREAD

**Issue by bocops** [andregasser/bigbone#456]
Some of our account-related methods, namely `verifyCredentials()` and `updateCredentials()`, return a `CredentialAccount` instead of an `Account` since the last snapshot release.

In our code, these entities are completely separate out of necessity (inheritance and data classes don't mix well), although Mastodon does not make this distinction. While some parts of the documentation specifically mention a "CredentialAccount" entity, others refer to the returned data as "an Account entity, with the source parameter included" (example: https://docs.joinmastodon.org/client/authorized/).

There doesn't seem to be any other way to retrieve the user's own account other than using `verifyCredentials()`, so I'd argue that this function is really meant to be used that way, and we either shouldn't return a completely different entity, or at least offer some way of getting an `Account` from the returned `CredentialAccount`. Potential ways to do this include:

- add a nullable `source` to `Account` and get rid of the separate `CedentialAccount`
- add a separate function (e.g. `getOwnAccount()`) that internally still calls `verifyCredentials()` but returns data as an `Account` (or possibly `Account?` if credentials can _not_ be verified)
- add a utility function that accepts a `CredentialAccount` and returns an `Account`

**Comment by PattaFeuFeu** [comment]
@bocops How do you feel about having `CredentialAccount` extend `Account`?

**Comment by bocops** [comment]
Extending data classes is not possible in Kotlin, so we couldn't for example declare `data class CredentialAccount(...) : Account(...)` and then later just cast to get the user's own account as a simple `Account`.

**Comment by PattaFeuFeu** [comment]
Ah, wow, I had totally forgotten about that again. 😕 

How do you feel about [this approach](https://stackoverflow.com/a/45756617/1791820)? Not great, but could be an option?

**Comment by bocops** [comment]
Similar to the above, I just found out that there's another Account detail we're not handling properly: currently, our method to retrieve accounts muted by the user, `MuteMethods.getMutes(range)` returns a `Pageable<Account>` in line with Mastodon's documentation of that endpoint.

However, Mastodon's documentation of the [Account entity](https://docs.joinmastodon.org/entities/Account/) describes a sub-entity `MutedAccount` with an additional attribute [mute_expires_at](url) that is not present for the generic `Account` entity.

Just like in the `CredentialAccount` case, we need to decide whether we want an additional entity, or if we just want to add these additional attributes as nullable to the main `Account`.

There are pros and cons in either case. Having just one `Account` entity for everything would spare us and our users some amount of casting between different sub-types - but at the same time may lead to the same instance of an `Account` entity having different interpretations. For example, `mute_expires_at=null` might mean that the account is indefinitely muted if it was retrieved from the mutes endpoint - but it would also be null for all accounts retrieved from other endpoints.

At the moment, I'm leaning towards creating different entities, but adding functions like `CredentialAccount.getAccount()` and `MutedAccount.getAccount()` that return standard `Account`.

**Comment by bocops** [comment]
Good idea. Turns out, we might already have that wrapper type in this specific instance. :)

`mute_expires_at` is supposed to be a datetime string, which we typically parse via our `PrecisionDateTime` interface with various implementing classes. So, what we _could_ do here, instead of creating a separate `MutedAccount`, is to add that attribute to `Account` along the following lines:

- `muteExpiresAt is Unavailable`: the account object was retrieved from an endpoint other than getMutes, and we don't know whether the account is muted
- `muteExpiresAt is ExactTime`: the account is muted, and the mute will expire at the given datetime
- `muteExpiresAt is Indefinite`: the account is muted, and the mute will never expire

In this case, `Indefinite` would be a new sub-class of either `ValidPrecisionDateTime` or `InvalidPrecisionDateTime` - whatever makes most sense.

At the moment, I'm not quite sure how to parse a null value differently depending on the endpoint we're retrieving it from, but perhaps there is a way that is not too cumbersome.

**Comment by bocops** [comment]
After some more consideration, I take back the above. We should not complicate one class containing some straight-forward information (is a datetime string exact, inexact or not available?) for a completely different purpose, which is the information whether an account is muted or not.

At the same time, I am still not completely happy with the idea of having three completely unrelated data classes for basically the same entity (an account on a Mastodon instance) with or without some extra details.

One option that hasn't been brought up yet, although we discussed inheritance earlier, is to simply make `Account` a proper `class` instead of `data class`. We would have to take care of `equals() `and `hashCode()` ourselves, but that shouldn't be too difficult considering we have a unique-per-instance `id` value.

**Comment by bocops** [comment]
The PR I just added fixes the missing `mute_expires_at `attribute when getting muted accounts by introducing a new `MutedAccount` entity separate from `Account`, similar to the already existing `CredentialAccount`.

This means a good amount of code duplication, but anything I could come up with in terms of inheritance just shifted this around instead of outright removing it. Unless I was looking at it the wrong way, the only sensible options seem to be either this, or moving all additional `CredentialAccount`/`MutedAccount` attributes into Account as optional attributes, which might come with its own set of problems as discussed above.

I'd be content with this either way, closing this or discussing further.

**Comment by PattaFeuFeu** [comment]
I’m a bit conflicted on which way to lean. I see the benefit of using regular classes instead of data classes for the cases where we want inheritance but dislike that we’d lose easy access to e.g. `copy` and that only some of our data classes would be actual `data class`es while a select few would be regular `class`es. 

Right now, I’m leaning towards just implementing the same class twice, thrice, or however often we need for the special cases with additional parameters (`CredentialAccount`, `MutedAccount`) and having a `toAccount` method to get the most basic variant. So basically what you did in #550.

I also like that you added the common type `Field` to the `Account` class to be reused by the other ones.

For structuring, it could make sense to have a separate package within `entity` to clarify that certain types belong together—`api.entity.account` which would then contain `Account`, `CredentialAccount`, `MutedAccount` for now.
That approach slightly falls apart because `entity.admin` also contains account-related entities but I think I’d be okay with the separation first by admin VS non-admin and then by common types.

**Comment by bocops** [comment]
Makes sense to discuss this aspect further. I changed the title of this issue to better reflect the remaining open discussion, now that the various entity variants are available in _some_ form.
